src/solidity/decodeInfo.js

```diff
@@ -0,0 +1,353 @@
+  */
+function StringType (type) {
+  return {
+    needsFreeStorageSlot: true,
+    storageBytes: 32,
+    typeName: 'string',
+    name: 'stringType'
+  }
+}
+
+/**
+  * ArrayType decode the given @arg type
+  *
+  * @param {String} type - type given by the AST
+  * @return {Object} returns decoded info about the current type: { needsFreeStorageSlot, storageBytes, typeName, decoder, arraySize, subArray}
+  */
+function ArrayType (type, stateDefinitions) {
+  var arraySize
+  var storageBytes
+
+  var underlyingType = extractUnderlyingType(type)
+
+  arraySize = extractArraySize(type)
+  var dimensions = extractDimensions(type)
+  type = underlyingType + dimensions.join('')
+
+  var subArrayType = type.substring(0, type.lastIndexOf('['))
+  var subArray = null
+  if (subArrayType.indexOf('[') !== -1) {
+    subArray = decode(subArrayType, stateDefinitions)
+  }
+
+  underlyingType = decode(underlyingType, stateDefinitions)
+  storageBytes = underlyingType.storageBytes
+
+  if (arraySize === 'dynamic') {
+    storageBytes = 32
+  } else {
+    if (subArray) {
+      storageBytes = subArray.storageBytes // size on storage of one item of the array
+    }
+    storageBytes = Math.ceil(storageBytes * arraySize) // size on storage of one the whole array
+  }
+
+  return {
+    needsFreeStorageSlot: true,
+    storageBytes: storageBytes,
+    typeName: type,
+    arraySize: arraySize,
+    subArray: subArray,
+    name: 'arrayType'
+  }
+}
+
+/**
+  * Enum decode the given @arg type
+  *
+  * @param {String} type - type given by the AST
+  * @return {Object} returns decoded info about the current type: { needsFreeStorageSlot, storageBytes, typeName, decoder}
+  */
+function Enum (type, stateDefinitions) {
+  var extracted = type.split(' ')
+  return {
+    needsFreeStorageSlot: false,
+    storageBytes: 1,
+    typeName: extracted[0] + ' ' + extracted[1],
+    enum: getEnum(type, stateDefinitions),
+    name: 'enum'
+  }
+}
+
+/**
+  * Struct decode the given @arg type
+  *
+  * @param {String} type - type given by the AST
+  * @return {Object} returns decoded info about the current type: { needsFreeStorageSlot, storageBytes, typeName, decoder, members}
+  */
+function Struct (type, stateDefinitions) {
+  var extracted = type.split(' ')
+  type = extracted[0] + ' ' + extracted[1]
+  var membersDetails = getStructMembers(type, stateDefinitions)
+  return {
+    needsFreeStorageSlot: true,
+    storageBytes: membersDetails.storageBytes,
+    typeName: type,
+    members: membersDetails.members,
+    name: 'struct'
+  }
+}
+
+/**
+  * retrieve enum declaration of the given @arg type
+  *
+  * @param {String} type - type given by the AST
+  * @param {Object} stateDefinitions  - all state declarations given by the AST (including struct and enum type declaration)
+  * @return {Array} - containing all value declaration of the current enum type
+  */
+function getEnum (type, stateDefinitions) {
```

Comment:
`type` contains the declared enum ( `enum enumDef` ) not the variable. So I think it's ok